inject/src/main/java/io/micronaut/context/AbstractBeanDefinition.java

```diff
@@ -2253,7 +2255,13 @@ private Object coerceCollectionToCorrectType(Class collectionType, Collection be
     private void addRequiredComponents(Argument... arguments) {
         if (arguments != null) {
             for (Argument argument : arguments) {
-                requiredComponents.add(argument.getType());
+                if (argument.isContainerType() || argument.isProvider()) {
+                    argument.getFirstTypeVariable()
+                            .map(Argument::getType)
+                            .ifPresent(requiredComponents::add);
+                } else {
+                    requiredComponents.add(argument.getType());
+                }
```

Comment:
The `isProvider` changes look ok, with regards to this change we are trying to eliminate all runtime evaluation in `AbstractBeanDefinition` with @dstepanov's PR so I would not like to add more here and this should be instead computed at build time by building on the work Denis has done.

Out of interest what is the use case? Since `getRequiredComponents` is very rarely used.

Reply:
This is used for shutdown behavior to ensure injected beans are shutdown after the bean they are injected into

Reply:
I would have to port it to my pr

Reply:
@dstepanov can you do that?

Reply:
Sure

Reply:
We can merge it so the tests are included